File: teamapps-client/projector-shaka-player/src/main/java/org/teamapps/projector/component/shakaplayer/ShakaPlayer.java
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package org.teamapps.projector.component.shakaplayer;

import org.teamapps.common.format.Color;
import org.teamapps.projector.component.AbstractComponent;
import org.teamapps.projector.component.ComponentConfig;
import org.teamapps.projector.event.ProjectorEvent;
import org.teamapps.projector.session.SessionContext;

public class ShakaPlayer extends AbstractComponent implements DtoShakaPlayerEventHandler {

	private final DtoShakaPlayerClientObjectChannel clientObjectChannel = new DtoShakaPlayerClientObjectChannel(getClientObjectChannel());

	public final ProjectorEvent<Void> onErrorLoading = new ProjectorEvent<>(clientObjectChannel::toggleErrorLoadingEvent);
	public final ProjectorEvent<DtoShakaManifest> onManifestLoaded = new ProjectorEvent<>(clientObjectChannel::toggleManifestLoadedEvent);
	public final ProjectorEvent<Long> onTimeUpdate = new ProjectorEvent<>(clientObjectChannel::toggleTimeUpdateEvent);
	public final ProjectorEvent<Void> onEnded = new ProjectorEvent<>(clientObjectChannel::toggleEndedEvent);

	public static void setDistinctManifestAudioTracksFixEnabled(boolean enabled) {
		SessionContext.current().sendStaticCommand(ShakaPlayer.class, DtoShakaPlayer.SetDistinctManifestAudioTracksFixEnabledCommand.CMD_NAME,
				new DtoShakaPlayer.SetDistinctManifestAudioTracksFixEnabledCommand(enabled));
	}

	private String hlsUrl;
	private String dashUrl;
	private String posterImageUrl;
	private PosterImageSize posterImageSize = PosterImageSize.COVER;
	private int timeUpdateEventThrottleMillis = 1000;
	private Color backgroundColor = Color.BLACK;
	private TrackLabelFormat trackLabelFormat = TrackLabelFormat.LABEL;
	private boolean videoDisabled = false;
	private String audioLanguage;

	private long timeMillis = 0;

	public ShakaPlayer() {
	}

	public ShakaPlayer(String hlsUrl, String dashUrl) {
		this.hlsUrl = hlsUrl;
		this.dashUrl = dashUrl;
	}

	@Override
	public ComponentConfig createConfig() {
		DtoShakaPlayer ui = new DtoShakaPlayer();
		mapAbstractConfigProperties(ui);
		ui.setHlsUrl(hlsUrl);
		ui.setDashUrl(dashUrl);
		ui.setPosterImageUrl(posterImageUrl);
		ui.setPosterImageSize(posterImageSize);
		ui.setTimeUpdateEventThrottleMillis(timeUpdateEventThrottleMillis);
		ui.setBackgroundColor(backgroundColor != null ? backgroundColor.toHtmlColorString() : null);
		ui.setTrackLabelFormat(trackLabelFormat);
		ui.setVideoDisabled(videoDisabled);
		ui.setTimeMillis(timeMillis);
		ui.setPreferredAudioLanguage(audioLanguage);
		return ui;
	}

	@Override
	public void handleErrorLoading() {
		onErrorLoading.fire(null);
	}

	@Override
	public void handleManifestLoaded(DtoShakaManifestWrapper manifest) {
		onManifestLoaded.fire(manifest.unwrap());
	}

	@Override
	public void handleTimeUpdate(long timeMillis) {
		onTimeUpdate.fire(timeMillis);
		this.timeMillis = timeMillis;
	}

	@Override
	public void handleEnded() {
		onEnded.fire();
	}

	public void setTime(long timeMillis) {
		this.timeMillis = timeMillis;
		clientObjectChannel.setTime(timeMillis);
	}

	public long getTime() {
		return timeMillis;
	}

	public void setUrls(String hlsUrl, String dashUrl) {
		this.timeMillis = 0;
		this.hlsUrl = hlsUrl;
		this.dashUrl = dashUrl;
		clientObjectChannel.setUrls(hlsUrl, dashUrl);
	}

	public String getHlsUrl() {
		return hlsUrl;
	}

	public void setHlsUrl(String hlsUrl) {
		setUrls(hlsUrl, dashUrl);
	}

	public String getDashUrl() {
		return dashUrl;
	}

	public void setDashUrl(String dashUrl) {
		setUrls(hlsUrl, dashUrl);
	}

	public String getPosterImageUrl() {
		return posterImageUrl;
	}

	public void setPosterImageUrl(String posterImageUrl) {
		this.posterImageUrl = posterImageUrl;
		clientObjectChannel.setPosterImageUrl(posterImageUrl);
	}

	public PosterImageSize getPosterImageSize() {
		return posterImageSize;
	}

	public void setPosterImageSize(PosterImageSize posterImageSize) {
		this.posterImageSize = posterImageSize;
		clientObjectChannel.setPosterImageSize(posterImageSize);
	}

	public int getTimeUpdateEventThrottleMillis() {
		return timeUpdateEventThrottleMillis;
	}

	public void setTimeUpdateEventThrottleMillis(int timeUpdateEventThrottleMillis) {
		if (clientObjectChannel.isRendered()) {
			throw new IllegalStateException("This cannot be set after rendering the player!");
		}
		this.timeUpdateEventThrottleMillis = timeUpdateEventThrottleMillis;
	}

	public Color getBackgroundColor() {
		return backgroundColor;
	}

	public void setBackgroundColor(Color backgroundColor) {
		this.backgroundColor = backgroundColor;
		clientObjectChannel.setBackgroundColor(backgroundColor.toHtmlColorString());
	}

	public TrackLabelFormat getTrackLabelFormat() {
		return trackLabelFormat;
	}

	public void setTrackLabelFormat(TrackLabelFormat trackLabelFormat) {
		if (clientObjectChannel.isRendered()) {
			throw new IllegalStateException("This cannot be set after rendering the player!");
		}
		this.trackLabelFormat = trackLabelFormat;
	}

	public boolean isVideoDisabled() {
		return videoDisabled;
	}

	public void setVideoDisabled(boolean videoDisabled) {
		this.videoDisabled = videoDisabled;
		clientObjectChannel.setVideoDisabled(videoDisabled);
	}

	public void selectAudioLanguage(String language) {
		selectAudioLanguage(language, null);
	}

	public void selectAudioLanguage(String language, String role) {
		this.audioLanguage = language;
		clientObjectChannel.selectAudioLanguage(language, role);
	}
}

File: client/projector-shaka-player/src/main/ts/ShakaPlayer.ts
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */

import "shaka-player";
import "shaka-player/dist/shaka-player.ui";
import {
	type DtoShakaManifest,
	type DtoShakaPlayer,
	type DtoShakaPlayer_EndedEvent,
	type DtoShakaPlayer_ErrorLoadingEvent,
	type DtoShakaPlayer_ManifestLoadedEvent, type DtoShakaPlayer_SkipClickedEvent,
	type DtoShakaPlayer_TimeUpdateEvent,
	type DtoShakaPlayerCommandHandler,
	type DtoShakaPlayerEventSource,
	type PosterImageSize,
	TrackLabelFormats,
} from "./generated";
import {
	AbstractComponent,
	executeAfterAttached,
	parseHtml,
	removeClassesByFunction,
	type ServerObjectChannel,
	ProjectorEvent,
	throttle, addDelegatedEventListener
} from "projector-client-object-api";
import Player = shaka.Player;
import UIConfiguration = shaka.extern.UIConfiguration;
import Overlay = shaka.ui.Overlay;
import ManifestConfiguration = shaka.extern.ManifestConfiguration;
import ManifestParser = shaka.extern.ManifestParser;
import PlayerInterface = shaka.extern.ManifestParser.PlayerInterface;
import TrackLabelFormat = shaka.ui.Overlay.TrackLabelFormat;
import Manifest = shaka.extern.Manifest;

import "./SkipBackButton";
import "./SkipForwardButton";

export class ShakaPlayer extends AbstractComponent<DtoShakaPlayer> implements DtoShakaPlayerCommandHandler, DtoShakaPlayerEventSource {

	public readonly onManifestLoaded: ProjectorEvent<DtoShakaPlayer_ManifestLoadedEvent> = new ProjectorEvent();
	public readonly onTimeUpdate: ProjectorEvent<DtoShakaPlayer_TimeUpdateEvent> = new ProjectorEvent();
	public readonly onEnded: ProjectorEvent<DtoShakaPlayer_EndedEvent> = new ProjectorEvent();
	public readonly onErrorLoading: ProjectorEvent<DtoShakaPlayer_ErrorLoadingEvent> = new ProjectorEvent();
	public readonly onSkipClicked: ProjectorEvent<DtoShakaPlayer_SkipClickedEvent> = new ProjectorEvent();


	private $componentWrapper: HTMLElement;
	private $video: HTMLVideoElement;
	private player: Player;

	private ui: Overlay;

	constructor(config: DtoShakaPlayer, serverObjectChannel: ServerObjectChannel) {
		super(config);

		console.log(config.hlsUrl, config.dashUrl, config.posterImageUrl)

		this.$componentWrapper = parseHtml(
			`<div class="ShakaPlayer">
				<div class="video-wrapper">
					<video id="video" ></video>
				</div>
			</div>`);

		this.$video = this.$componentWrapper.querySelector(":scope video");
		this.$video.addEventListener("timeupdate", throttle(e => this.onTimeUpdate.fire({timeMillis: this.$video.currentTime * 1000}), config.timeUpdateEventThrottleMillis))
		this.$video.addEventListener("ended", e => this.onEnded.fire({}))

		this.setPosterImageUrl(config.posterImageUrl);
		this.setPosterImageSize(config.posterImageSize);
		this.setBackgroundColor(config.backgroundColor);

		this.displayedDeferredExecutor.invokeWhenReady(() => {
			this.player = new shaka.Player(this.$video);
			this.reconfigurePlayer();
			this.player.addEventListener('error', () => this.onErrorLoading.fire({}));
			this.player.addEventListener('streaming', () => {
				this.reconfigureUi();
				this.jumpTo(this.config.timeMillis);
				this.onManifestLoaded.fire({manifest: this.createUiManifest(this.player.getManifest())});
			});

			this.ui = new shaka.ui.Overlay(this.player, this.$componentWrapper, this.$video);
			const uiConfig: Partial<UIConfiguration> = {
				addBigPlayButton: config.bigPlayButtonEnabled,
				controlPanelElements: config.controlPanelElements,
				doubleClickForFullscreen: true,
				enableFullscreenOnRotation: false,
				enableKeyboardPlaybackControls: true,
				fadeDelay: config.controlFadeDelaySeconds,
				forceLandscapeOnFullscreen: true,
				overflowMenuButtons: this.getOverflowMenuButtonsConfig(),
				seekBarColors: undefined,
				showUnbufferedStart: false,
				trackLabelFormat: config.trackLabelFormat == TrackLabelFormats.LABEL ? TrackLabelFormat.LABEL
					: config.trackLabelFormat == TrackLabelFormats.LANGUAGE ? TrackLabelFormat.LANGUAGE
						: config.trackLabelFormat == TrackLabelFormats.ROLE ? TrackLabelFormat.ROLE
							: config.trackLabelFormat == TrackLabelFormats.LANGUAGE_ROLE ? TrackLabelFormat.LANGUAGE_ROLE
								: undefined,
				volumeBarColors: undefined,
				addSeekBar: true
			};
			this.ui.configure(uiConfig as UIConfiguration);

			addDelegatedEventListener(this.ui.getControls().getControlsContainer(), '.shaka-skip-button', 'click', (el, ev) => {
				this.onSkipClicked.fire({
					forward: el.classList.contains('shaka-skip-forward-button'),
					playbackTimeMillis: this.player.getMediaElement().currentTime * 1000
				})
			}, {capture: true})
		});

		this.setUrls(config.hlsUrl, config.dashUrl);
	}

	setBigPlayButtonEnabled(bigPlayButtonEnabled: boolean) {
		throw new Error("Method not implemented.");
    }
    setControlFadeDelaySeconds(controlFadeDelaySeconds: number) {
        throw new Error("Method not implemented.");
    }

	private createUiManifest(manifest: Manifest): DtoShakaManifest {
		return {
			variants: manifest.variants.map(v => {
				return {
					id: v.id,
					audio: {
						id: v.audio.id,
						originalId: v.audio.originalId,
						mimeType: v.audio.mimeType,
						codecs: v.audio.codecs,
						bandwidth: v.audio.bandwidth,
						label: v.audio.label,
						roles: v.audio.roles,
						language: v.audio.language,
						channelsCount: v.audio.channelsCount,
						audioSamplingRate: v.audio.audioSamplingRate
					},
					video: {
						id: v.video.id,
						originalId: v.video.originalId,
						mimeType: v.video.mimeType,
						codecs: v.video.codecs,
						bandwidth: v.video.bandwidth,
						label: v.video.label,
						roles: v.video.roles,
						frameRate: v.video.frameRate,
						pixelAspectRatio: v.video.pixelAspectRatio,
						width: v.video.width,
						height: v.video.height
					},
					bandwidth: v.bandwidth
				}
			})
		};
	}

	private reconfigurePlayer() {
		this.player.configure({
			manifest: {
				disableVideo: this.config.videoDisabled // will advice the manifest parser to set all videos to null. Note that the quality selection button must be removed!
			},
			preferredAudioLanguage: this.config.preferredAudioLanguage
		} );
	}

	private reconfigureUi() {
		this.ui.configure({
			overflowMenuButtons: this.getOverflowMenuButtonsConfig()
		} as Partial<UIConfiguration>);
	}

	private getOverflowMenuButtonsConfig() {
		let manifest = this.player.getManifest();
		const overflowButtons = [
			"playback_rate",
			"captions",
			// "picture_in_picture",
		];
		if (manifest != null && manifest.variants.some(v => v.video != null)) {
			overflowButtons.unshift("quality");
		}
		if (manifest != null && manifest.variants.some(v => v.language && v.language !== "und")) {
			overflowButtons.unshift("language");
		}
		return overflowButtons;
	}

	public doGetMainElement(): HTMLElement {
		return this.$componentWrapper;
	}


	public play(): any {
		this.$video.play();
	}

	public pause(): any {
		this.$video.pause();
	}

	public jumpTo(timeMillis: number) {
		this.$video.currentTime = timeMillis / 1000;
	}

	public selectAudioLanguage(language: string, role: string): any {
		this.config.preferredAudioLanguage = language;
		this.player.selectAudioLanguage(language, role);
	}

	setPosterImageUrl(posterImageUrl: string) {
		this.config.posterImageUrl = posterImageUrl;
		this.$video.poster = posterImageUrl;
	}
	setPosterImageSize(posterImageSize: PosterImageSize) {
		this.config.posterImageSize = posterImageSize;
		removeClassesByFunction(this.$video.classList, className => className.startsWith("poster-"));
		this.$video.classList.add(`poster-${this.config.posterImageSize}`)
	}
	setBackgroundColor(backgroundColor: string) {
		this.$componentWrapper.style.backgroundColor = backgroundColor;
	}
	setVideoDisabled(videoDisabled: boolean) {
		this.config.videoDisabled = videoDisabled;
		this.reconfigurePlayer();
	}

	public onResize(): void {

	}

	public destroy(): void {
	}

	@executeAfterAttached(true)
	async setUrls(hlsUrl: string, dashUrl: string): Promise<void> {
		const support = await shaka.Player.probeSupport();
		let url = support.manifest.mpd && dashUrl ? dashUrl : hlsUrl;
		console.log(url)
		try {
			await this.player.load(url);
			console.log(this.player.getConfiguration().preferredAudioLanguage)
		} catch (e) {
			console.error(e);
			this.onErrorLoading.fire({});
		}
	}

	static setDistinctManifestAudioTracksFixEnabled(enabled: boolean) {
		const dashParserFactory = enabled ? () => new DistinctAudioTracksManifestParserDecorator(new shaka.dash.DashParser()) : () => new shaka.dash.DashParser();
		shaka.media.ManifestParser.registerParserByExtension('mpd', dashParserFactory);
		shaka.media.ManifestParser.registerParserByMime('application/dash+xml', dashParserFactory);
		shaka.media.ManifestParser.registerParserByMime('video/vnd.mpeg.dash.mpd', dashParserFactory);
		const hlsParserFactory = enabled ? () => new DistinctAudioTracksManifestParserDecorator(new shaka.hls.HlsParser()) : () => new shaka.hls.HlsParser();
		shaka.media.ManifestParser.registerParserByExtension('m3u8', hlsParserFactory);
		shaka.media.ManifestParser.registerParserByMime('application/x-mpegurl', hlsParserFactory);
		shaka.media.ManifestParser.registerParserByMime('application/vnd.apple.mpegurl', hlsParserFactory);
	}
}

class DistinctAudioTracksManifestParserDecorator implements ManifestParser {
	constructor(private delegate: ManifestParser) {
	}

	configure(config: ManifestConfiguration) {
		return this.delegate.configure(config);
	}

	async start(uri: string, playerInterface: PlayerInterface) {
		let manifest = await this.delegate.start(uri, playerInterface);
		// Add a unique role for each audio track.
		// manifest.variants.forEach((variant, i) => variant.language = variant.language + i)
		manifest.variants.forEach((variant, i) => variant.audio.roles.push("role" + i));
		return manifest;
	}

	async stop() {
		return this.delegate.stop();
	}

	onExpirationUpdated(sessionId: string, expiration: number): any {
		return this.delegate.onExpirationUpdated(sessionId, expiration);
	}

	update(): any {
		return this.delegate.update();
	}

	banLocation(location: string): void {
		return this.delegate.banLocation(location);
	}

	onInitialVariantChosen(variant: shaka.extern.Variant): void {
		return this.delegate.onInitialVariantChosen(variant);
	}

	setMediaElement(mediaElement: HTMLMediaElement): void {
		return this.delegate.setMediaElement(mediaElement);
	}
}

shaka.polyfill.installAll();


